### DISCUSSION TITLE
NUV prior with Delta node

### DISCUSSION THREAD

**Discussion by Chengfeng-Jia** [ReactiveBayes/RxInfer.jl#311] [Q&A] [answered]
Hey guys,
Recently, I experimented with the NUV prior and nonlinear dynamics using some previous code. My goal was to determine the optimal control policy $u_{\{k,t\}}$  for obstacle avoidance. 
	
#  
		x[k,1] ~ MvNormalMeanCovariance(zeros(5), 1e-3I)
	
		for t in 1:nr_steps
			
			# prior on controls
			u[k,t] ~ MvNormalMeanCovariance(zeros(1), 1e-4I)
			
			# Nonlinear state transition
			x[k,t+1]  ~ get_matrix_B(x[k,t],u[k,t]) 
	
			# observation model
			y[k,t] ~ C * x[k,t+1]

			# environmental distance
			zσ2[k,t] ~ GammaShapeRate(3/2, γ^2/2)
			z[k,t] ~ g(y[k,t])
			z[k,t] ~ Halfspace(0, zσ2[k,t], γ)

		end

		# goal priors 
		goals[1,k] ~ MvNormalMeanCovariance(x[k,1], 1e-5I)
		goals[2,k] ~ MvNormalMeanCovariance(x[k,end], aa)

While I obtained a relatively good path for the posterior of $x_{k,t}$, there is a significant deviation between this path and the corresponding inferred control $u_{k,t}$. 
To better illustrate what I said, I drew this figure:
![d20e945e-942a-44f9-b3d6-0adaa749d0cb](https://github.com/ReactiveBayes/RxInfer.jl/assets/97285031/7d77bb86-3a14-4bf0-a273-8ee6cde07994)

The red dashed line represents the posterior of $x_{k,t}$. The blue line indicates the executed path given the initial position $y_{k,1}$ and the control policy $u_{\{k,t\}}$.

Since the NUV prior serves as a constraint on the observations/state, could this affect the "accuracy" of the Delta node in RxInfer?

**Comment by albertpod** [comment]
Hi @Chengfeng-Jia! Not so sure what's causing it.

@bartvanerp, any idea?

**Reply by bartvanerp** [reply]
Hi @Chengfeng-Jia,

Do you also observe the same behaviour without the NUV priors?

What you could already check is whether it helps to apply the NUV prior directly on the latent state (`g(y[])`).

**Comment by Chengfeng-Jia** [comment, marked as answer]
Hi @albertpod and @bartvanerp, thanks for your reply and suggestions.
I removed the NUV prior and found that the issue lies in the linearization process itself.

Although the goal prior has a tiny covariance and can yield a relatively good result for the state $x$ (path) to reach the goal, the low accuracy of the linearization prevents it from providing the corresponding control $u$. 

Therefore, when I replaced the complex nonlinear transition with a simpler nonlinear transition, it worked.

**Reply by bartvanerp** [reply]
Hi @Chengfeng-Jia,

Good that you found what is causing the issue. We are working on creating more reliable approximations to prevent these kinds of problems from arising in the future :)
